aten/src/ATen/native/cpu/UpSampleKernel.cpp

```diff
@@ -965,22 +966,29 @@ struct HelperInterpCubic : public HelperInterpBase {
 
     AT_DISPATCH_FLOATING_TYPES_AND(
       ScalarType::BFloat16, scalar_type, "compute_indices_weights_cubic", [&] {
-        scalar_t scale = area_pixel_compute_scale<scalar_t>(input_size, output_size, align_corners, opt_scale);
+        using opmath_t = at::opmath_type<scalar_t>;
+        opmath_t scale = area_pixel_compute_scale<opmath_t>(input_size, output_size, align_corners, opt_scale);
 
         int64_t input_index;
         int64_t zero = static_cast<int64_t>(0);
         // NOLINTNEXTLINE(cppcoreguidelines-avoid-c-arrays,modernize-avoid-c-arrays)
-        scalar_t coeffs[4];
+        opmath_t coeffs[4];
 
         int64_t * idx_ptr;
         scalar_t * wt_ptr;
-        using opmath_t = at::opmath_type<scalar_t>;
         for (const auto i : c10::irange(output_size)) {
           const auto real_input_index =
               area_pixel_compute_source_index<opmath_t>(
                   scale, i, align_corners, /*cubic=*/true);
-          input_index = static_cast<int64_t>(floorf(real_input_index));
-          get_cubic_upsample_coefficients<scalar_t>(coeffs, real_input_index - input_index);
+          // when `real_input_index` becomes larger than the range the floating point
+          // type can accurately represent, the type casting to `int64_t` might exceed
+          // `input_size - 1`. So we guard it with `std::min` below.
+          input_index = std::min(static_cast<int64_t>(floorf(real_input_index)), input_size - 1);
+          auto lambda = std::min(
+            std::max(real_input_index - input_index, static_cast<opmath_t>(0)),
+            static_cast<opmath_t>(1)
+          );
```

Comment:
Seems there are quite a few duplicate code like this. Can we factor out a util function to dedup?

Reply:
factor out a util function `guard_index_and_lambda`.